Fixes #9028 - Make priority unique

boaz0 opened the pull request:
@lzap would you mind looking at this change. In addition, I didn't write any unit-tests, because I am not sure how to run the tests.

Thanks.

boaz0 committed:
Fixes 9028

Right now, two discovery rules may have the same priority.
This however may yield unexpected results or more likely
is not sustainable when you have many discovery rules.

In order to fix this, it is decided to make priority unique.
By doing so we gain:
 (a) Make it easy to reason which discovery rule will be
     executed
 (b) Change the search query to something simple and readable:
     DiscoveryRule.where(:enabled => true).reorder(:priority)

Signed-off-by: Boaz Shuster <ripcurld.github@gmail.com>

lzap commented on db/migrate/20180412124505_add_priority_score_to_discovery_rules.rb:
```diff
@@ -0,0 +1,23 @@
+  end
+
+  def up
+    score = DiscoveryRules.distinct.order("priority asc").pluck(:priority).first
```
This is the query how we search discovery rules at the moment, can you do the very same query so we know for sure there are no regressions caused by incorrect oder after migration?

    DiscoveryRule.where(:enabled => true).reorder(:priority, :created_at)

Add `unscoped` as well to make sure we get them all. Also use batch processing, so I'd expect something like:

    DiscoveryRule.unscoped.where(:enabled => true).reorder(:priority, :created_at).find_each

boaz0 replied:
Thanks @lzap for the fast response. Sorry it took me too long.

I didn't add `where(:enabled => true)` to the migration script because we want to go over all the rules and verify that both enabled and disabled rules have different priority.

If you believe that's unnecessary, let me know.

Thanks again :grin:

lzap commented on db/migrate/20180412124505_add_priority_score_to_discovery_rules.rb:
```diff
@@ -0,0 +1,23 @@
+    rules = DiscoveryRules.order("priority asc, name asc")
+    rules.each do |rule|
+      rule.update_attributes!(:priority => score)
+      score += 1
```
Can you simplify that and set initial score to `100` and then add `100` every loop so there is enough room for additional entries? I think we can ignore current priority numbers totally, they are not important. What's important is the order we must kept.

lzap commented on db/migrate/20180412124505_add_priority_score_to_discovery_rules.rb:
```diff
@@ -0,0 +1,23 @@
+      score += 1
+      say "#{rule.name} priority is #{rule.priority}"
+    end
+    change_column :discovery_rules, :priority, :integer, unique: true, default: 0
```
The `unique` option does not seem to be right. We actually want this column to be always present with no default: `null: false` - note the default is no longer here because we want it to fail when there is no priority. Then add index to guarantee consistency and faster search queries:

    add_index(:discovery_rules, :priority, unique: true)

lzap reviewed (changes requested):
Hey, very nice thanks. I have couple of comments in the migration, but most importantly we can now change this query in `find_discovery_rule` method to take advantage of this new database layout. Instead of:

DiscoveryRule.where(:enabled => true).reorder(:priority, :created_at)

we can simply do this:

DiscoveryRule.where(:enabled => true).reorder(:priority)

lzap commented:
We are all good, can you rebase the commit message to Fixes #XXXX where XXXX is issue number from our tracker: http://projects.theforeman.org/projects/discovery/issues/new (if you haven' created one please create new one)

boaz0 commented:
Thanks a lot :smile:  @lzap I have removed one of the tests as it's not relevant anymore due to this change and also modified two of the tests because they had the same priority.

I tried to execute the tests on my machine using `bundle exec rake test:discovery` but it fails on all the tests. I wish I knew how to execute them properly, though.

Anyway, please forgive me if the tests fail, it will take me time to figure out how to pass. :trollface: 

Thanks. :trophy:

lzap reviewed (approved):
Tests are green locally, looks good. Nice job there! I will squash and merge once it's green on CI.

lzap commented:
Thanks a bunch! Nice patch.

boaz0 commented:
@lzap it was a pleasure working on this with you.
Thank you so much! :raised_hands:
